File: lcc_core/fit.py
from __future__ import annotations

import re
import shlex
import subprocess
from pathlib import Path
from typing import Any

from .estimates import estimate_tokens_per_second
from .hardware import detect_system_hardware
from .llama_args import normalize_gpu_layers
from .server_manager import prepare_launch_command


MEMORY_RE = re.compile(
    r"(?im)^(?P<device>(?:CUDA|MTL|METAL|ROCM|HIP|VULKAN|VK|SYCL|GPU)\d*)\s+"
    r"(?P<model>\d+)\s+(?P<context>\d+)\s+(?P<compute>\d+)"
)
FREE_RE = re.compile(r"projected to use (?P<used>\d+) MiB.* vs\. (?P<free>\d+) MiB", re.IGNORECASE)
LEAVE_RE = re.compile(r"will leave (?P<remaining>\d+).*?MiB", re.IGNORECASE)

INT_FLAGS = {
    "-c": "ctx_size",
    "--ctx-size": "ctx_size",
    "-t": "threads",
    "--threads": "threads",
    "-tb": "threads_batch",
    "--threads-batch": "threads_batch",
    "-b": "batch_size",
    "--batch-size": "batch_size",
    "-ub": "ubatch_size",
    "--ubatch-size": "ubatch_size",
    "--repeat-last-n": "repeat_last_n",
    "-s": "seed",
    "--seed": "seed",
    "-n": "n_predict",
    "--predict": "n_predict",
    "--n-predict": "n_predict",
}
FLOAT_FLAGS = {
    "--temp": "temperature",
    "--temperature": "temperature",
    "--top-p": "top_p",
    "--min-p": "min_p",
    "--repeat-penalty": "repeat_penalty",
    "--presence-penalty": "presence_penalty",
    "--frequency-penalty": "frequency_penalty",
}
STRING_FLAGS = {
    "-ctk": "cache_type_k",
    "--cache-type-k": "cache_type_k",
    "-ctv": "cache_type_v",
    "--cache-type-v": "cache_type_v",
}
GPU_LAYER_FLAGS = {"-ngl", "--gpu-layers", "--n-gpu-layers"}
BOOL_FLAGS = {"-fa": "flash_attn", "--flash-attn": "flash_attn", "--reasoning": "reasoning"}
BOOL_FLAGS.update({"-kvo": "kv_offload", "--kv-offload": "kv_offload", "--op-offload": "op_offload"})
NEG_BOOL_FLAGS = {
    "-nkvo": "kv_offload",
    "--no-kv-offload": "kv_offload",
    "--no-op-offload": "op_offload",
}
INT_OR_ZERO_FLAGS = {"--top-k": "top_k"}
FIT_APPLY_KEYS = {
    "ctx_size",
    "threads",
    "threads_batch",
    "batch_size",
    "ubatch_size",
    "gpu_layers",
    "cache_type_k",
    "cache_type_v",
    "temperature",
    "top_k",
    "top_p",
    "min_p",
    "repeat_penalty",
    "repeat_last_n",
    "presence_penalty",
    "frequency_penalty",
    "seed",
    "n_predict",
    "flash_attn",
    "reasoning",
    "kv_offload",
    "op_offload",
}


def build_fit_args(fit_binary: str, model_path: str, params: dict[str, Any], target_mib: int = 1024) -> list[str]:
    gpu_layers = normalize_gpu_layers(params.get("gpu_layers"))
    if gpu_layers is None:
        gpu_layers = 999
    gpu_layers_arg = "-2" if gpu_layers >= 999 else str(gpu_layers)
    args = [
        fit_binary,
        "-m",
        model_path,
        "-c",
        str(int(params.get("ctx_size", 4096))),
        "-ngl",
        gpu_layers_arg,
        "-t",
        str(int(params.get("threads", 4))),
        "-tb",
        str(int(params.get("threads_batch", params.get("threads", 4)))),
        "-ctk",
        str(params.get("cache_type_k", "q4_0")),
        "-ctv",
        str(params.get("cache_type_v", "q4_0")),
        "-b",
        str(int(params.get("batch_size", 512))),
        "-ub",
        str(int(params.get("ubatch_size", 512))),
        "-fit",
        "on",
        "-fitp",
        "on",
        "-fitt",
        str(int(target_mib)),
    ]
    if params.get("flash_attn", True):
        args.extend(["-fa", "on"])
    args.append("-kvo" if params.get("kv_offload", True) else "-nkvo")
    args.append("--op-offload" if params.get("op_offload", True) else "--no-op-offload")
    if "temperature" in params:
        args.extend(["--temp", str(params["temperature"])])
    if "top_k" in params:
        args.extend(["--top-k", str(params["top_k"])])
    if "top_p" in params:
        args.extend(["--top-p", str(params["top_p"])])
    if "min_p" in params:
        args.extend(["--min-p", str(params["min_p"])])
    if "repeat_penalty" in params:
        args.extend(["--repeat-penalty", str(params["repeat_penalty"])])
    if "repeat_last_n" in params:
        args.extend(["--repeat-last-n", str(params["repeat_last_n"])])
    if "presence_penalty" in params:
        args.extend(["--presence-penalty", str(params["presence_penalty"])])
    if "frequency_penalty" in params:
        args.extend(["--frequency-penalty", str(params["frequency_penalty"])])
    if "seed" in params:
        args.extend(["-s", str(params["seed"])])
    if "n_predict" in params:
        args.extend(["--predict", str(params["n_predict"])])
    if "reasoning" in params and params["reasoning"]:
        args.append("--reasoning")
    return args


def _find_fitted_args(stdout: str) -> str | None:
    for raw_line in stdout.splitlines():
        line = raw_line.strip()
        if not line:
            continue
        has_context = " -c " in f" {line} " or " --ctx-size " in f" {line} "
        has_layers = any(f" {flag} " in f" {line} " for flag in GPU_LAYER_FLAGS)
        if not has_context or not has_layers:
            continue
        # Slice from the earliest flag so a leading prose prefix is dropped but
        # no flag is lost if -ngl happens to precede -c in the tool's output.
        markers = ["-c ", "--ctx-size "] + [f"{flag} " for flag in GPU_LAYER_FLAGS]
        positions = [pos for pos in (line.find(marker) for marker in markers) if pos >= 0]
        if positions:
            return line[min(positions):]
    return None


def _tokens(command_line: str) -> list[str]:
    try:
        return shlex.split(command_line, posix=False)
    except ValueError:
        return command_line.split()


def _bool_value(value: str | None) -> bool:
    if value is None:
        return True
    return value.lower() not in {"0", "false", "off", "no"}


def parse_fitted_args(command_line: str) -> dict[str, Any]:
    suggestions: dict[str, Any] = {"fitted_args": command_line.strip()}
    tokens = _tokens(command_line)
    idx = 0
    while idx < len(tokens):
        token = tokens[idx]
        next_value = tokens[idx + 1] if idx + 1 < len(tokens) else None
        if token in GPU_LAYER_FLAGS and next_value is not None:
            suggestions["gpu_layers"] = 999 if next_value in {"-2", "all", "auto"} else int(next_value)
            idx += 2
            continue
        if token in INT_FLAGS and next_value is not None:
            suggestions[INT_FLAGS[token]] = int(float(next_value))
            idx += 2
            continue
        if token in INT_OR_ZERO_FLAGS and next_value is not None:
            suggestions[INT_OR_ZERO_FLAGS[token]] = int(float(next_value))
            idx += 2
            continue
        if token in FLOAT_FLAGS and next_value is not None:
            suggestions[FLOAT_FLAGS[token]] = float(next_value)
            idx += 2
            continue
        if token in STRING_FLAGS and next_value is not None:
            suggestions[STRING_FLAGS[token]] = str(next_value)
            idx += 2
            continue
        if token in BOOL_FLAGS:
            consumes_value = next_value is not None and not next_value.startswith("-")
            suggestions[BOOL_FLAGS[token]] = _bool_value(next_value if consumes_value else None)
            idx += 2 if consumes_value else 1
            continue
        if token in NEG_BOOL_FLAGS:
            suggestions[NEG_BOOL_FLAGS[token]] = False
            idx += 1
            continue
        idx += 1
    return suggestions


def apply_fit_suggestions(params: dict[str, Any], suggestions: dict[str, Any], target_mib: int | None = None) -> dict[str, Any]:
    applied = dict(params)
    for key in FIT_APPLY_KEYS:
        if key in suggestions:
            applied[key] = suggestions[key]
    if target_mib is not None:
        applied["fit_target_mib"] = int(target_mib)
    if "headroom_mib" in suggestions:
        applied["fit_headroom_mib"] = suggestions["headroom_mib"]
    if "cuda_memory_mib" in suggestions:
        applied["fit_cuda_memory_mib"] = suggestions["cuda_memory_mib"]
    return applied


def parse_fit_output(stdout: str, stderr: str = "") -> dict[str, Any]:
    combined = "\n".join(part for part in [stdout, stderr] if part)
    fitted_line = _find_fitted_args(stdout)
    memory = MEMORY_RE.search(combined)
    free_match = FREE_RE.search(combined)
    leave_match = LEAVE_RE.search(combined)
    suggestions: dict[str, Any] = {}
    notes: list[str] = []

    if fitted_line:
        suggestions.update(parse_fitted_args(fitted_line))
        notes.append("llama-fit-params returned fitted CLI arguments.")

    if memory:
        model_mib = int(memory.group("model"))
        context_mib = int(memory.group("context"))
        compute_mib = int(memory.group("compute"))
        suggestions["cuda_memory_mib"] = {
            "model": model_mib,
            "context": context_mib,
            "compute": compute_mib,
            "projected": model_mib + context_mib + compute_mib,
        }
        notes.append(f"Parsed {memory.group('device')} memory estimate from llama-fit-params output.")

    if free_match:
        used = int(free_match.group("used"))
        free = int(free_match.group("free"))
        suggestions["projected_device_mib"] = used
        suggestions["free_device_mib"] = free
        suggestions["headroom_mib"] = max(free - used, 0)
    if leave_match:
        suggestions["headroom_mib"] = int(leave_match.group("remaining"))

    if not suggestions:
        notes.append("No structured fit suggestion could be parsed; review stdout/stderr.")

    return {"suggestions": suggestions, "notes": notes}


def run_fit_test(
    mode: str,
    project_root: str | Path | None = None,
    model_dirs: list[str | Path] | None = None,
    overrides: dict[str, Any] | None = None,
    target_mib: int = 1024,
    timeout_seconds: int = 180,
) -> dict[str, Any]:
    prepared = prepare_launch_command(mode, project_root=project_root, model_dirs=model_dirs, overrides=overrides)
    if not prepared.get("success"):
        return prepared

    fit_binary = prepared.get("environment", {}).get("details", {}).get("llama_fit_params")
    if not fit_binary or not Path(fit_binary).is_file():
        return {"success": False, "error": "llama-fit-params was not found.", "prepared": prepared}

    model_path = prepared["profile"]["model"]["path"]
    args = build_fit_args(fit_binary, model_path, prepared["params"], target_mib)
    cwd = str(Path(fit_binary).parent)
    try:
        result = subprocess.run(
            args,
            cwd=cwd,
            capture_output=True,
            text=True,
            timeout=int(timeout_seconds),
            check=False,
        )
    except Exception as exc:
        return {"success": False, "error": str(exc), "prepared": prepared, "argv": args}

    parsed = parse_fit_output(result.stdout, result.stderr)
    applied_params = apply_fit_suggestions(prepared["params"], parsed["suggestions"], target_mib)
    hardware = detect_system_hardware()
    speed_estimate = estimate_tokens_per_second(
        applied_params,
        prepared.get("profile", {}).get("model"),
        hardware,
    )
    return {
        "success": result.returncode == 0,
        "returncode": result.returncode,
        "argv": args,
        "command_line": subprocess.list2cmdline(args),
        "stdout": result.stdout,
        "stderr": result.stderr,
        "prepared": prepared,
        "applied_params": applied_params,
        "hardware": hardware,
        "speed_estimate": speed_estimate,
        **parsed,
    }

File: lcc_api/static/js/panels/fit.js
// Fit panel.

import { getSelectedProfile, profileLabel, renderProfiles, setSelectedProfileMode } from './profiles.js';
import { applyFitResultParams, collectOverrides, markAppliedFields, primaryGpu, renderParameters, saveCurrentOverrides, selectedMode, setParamOverrides } from './parameters.js';
import { setModelNote } from './models.js';
import { showLogPreview } from './logs.js';
import { $, escapeHtml, formatMib, formatNumber, hasOwn } from '../util.js';
import { renderTuneSummary, shouldAutoApplyTune } from '../tune.js';
import { state } from '../state.js';
import { refresh } from '../refresh.js';
import { fitItem, fitStatusLabel } from '../format.js';
import { confirmAction, toast, withBusy } from '../feedback.js';
import { api } from '../api.js';

export function renderTpsEstimate(estimate) {
  const value = $('#tps-estimate');
  const detail = $('#tps-detail');
  const card = $('#speed-estimate-card');
  if (card) card.classList.add('updating');
  if (!estimate) {
    value.textContent = '-';
    detail.textContent = 'Waiting for model and hardware details.';
    if (card) setTimeout(() => card.classList.remove('updating'), 400);
    return;
  }
  value.textContent = `${estimate.estimate_tps} tok/s`;
  detail.textContent = `${estimate.low_tps}-${estimate.high_tps} tok/s range, ${estimate.confidence} confidence`;
  if (card) setTimeout(() => card.classList.remove('updating'), 400);
}

export function renderMeasuredTps(tokensPerSecond, elapsed) {
  const value = $('#tps-estimate');
  const detail = $('#tps-detail');
  if (tokensPerSecond === undefined || tokensPerSecond === null || tokensPerSecond === 0) {
    value.textContent = '-';
    detail.textContent = 'Waiting for model and hardware details.';
    return;
  }
  value.textContent = `${tokensPerSecond} tok/s (measured)`;
  detail.textContent = `Benchmark result: ${elapsed}s elapsed`;
  state.measuredTps = tokensPerSecond;
  state.measuredElapsed = elapsed;
  state.lastBenchmarkKey = estimateKey(selectedMode() || '', collectOverrides());
}

export function shouldShowMeasuredTps() {
  if (!state.measuredTps || !state.lastBenchmarkKey) return false;
  const currentKey = estimateKey(selectedMode() || '', collectOverrides());
  return currentKey === state.lastBenchmarkKey;
}

export function renderFitEstimate(fit) {
  const value = $('#fit-estimate');
  const detail = $('#fit-detail');
  const card = $('#fit-estimate-card');
  card.classList.remove('status-good', 'status-tight', 'status-near-limit');
  if (card) card.classList.add('updating');
  if (!fit) {
    value.textContent = '-';
    detail.textContent = 'Waiting for model and hardware details.';
    if (card) setTimeout(() => card.classList.remove('updating'), 400);
    return;
  }
  const estimated = fit.estimated || {};
  const status = fit.status || 'unknown';
  value.textContent = fit.label || fitStatusLabel(status);
  if (status === 'good') card.classList.add('status-good');
  if (status === 'tight') card.classList.add('status-tight');
  if (status === 'near_limit') card.classList.add('status-near-limit');
  const details = [
    `VRAM use ${formatMib(estimated.accelerator_used_mib)}`,
    `headroom ${formatMib(estimated.accelerator_headroom_mib)}`,
  ];
  if (fit.uses_ram_offload) {
    details.push(`RAM use ${formatMib(estimated.ram_used_mib)}`);
  }
  detail.textContent = details.join(' · ');
  if (card) setTimeout(() => card.classList.remove('updating'), 400);
}

export function clearMeasuredTps() {
  state.measuredTps = null;
  state.measuredElapsed = null;
  state.lastBenchmarkKey = '';
}

export function renderEstimatePending(message = 'Estimating launch...') {
  $('#fit-estimate').textContent = '-';
  $('#fit-detail').textContent = message;
  $('#tps-estimate').textContent = '-';
  $('#tps-detail').textContent = message;
  clearMeasuredTps();
}

export function estimateKey(mode, overrides) {
  return JSON.stringify({
    mode,
    overrides,
    gpu: primaryGpu()?.name || '',
    vram: primaryGpu()?.vram_total_bytes || 0,
    ram: state.hardware?.memory?.total_bytes || 0,
    cpu: state.hardware?.cpu?.logical_cores || 0,
  });
}

export function scheduleTpsEstimate(delay = 350) {
  window.clearTimeout(scheduleTpsEstimate.timer);
  scheduleTpsEstimate.timer = window.setTimeout(updateTpsEstimate, delay);
}

export async function updateTpsEstimate() {
  const mode = selectedMode();
  if (!mode) {
    renderTpsEstimate(null);
    renderFitEstimate(null);
    return;
  }
  const overrides = collectOverrides();
  const key = estimateKey(mode, overrides);
  if (state.lastEstimateKey === key) return;
  state.lastEstimateKey = key;
  renderEstimatePending();
  try {
    const result = await api('/api/estimate/launch', {
      method: 'POST',
      body: JSON.stringify({ mode, overrides }),
    });
    renderTpsEstimate(result.speed_estimate);
    renderFitEstimate(result.fit_status);
  } catch (error) {
    $('#fit-estimate').textContent = '-';
    $('#fit-detail').textContent = `Fit unavailable: ${error.message}`;
    $('#tps-estimate').textContent = '-';
    $('#tps-detail').textContent = `Estimate unavailable: ${error.message}`;
  }
}

export function parsedFitAccepted(suggestions) {
  return Object.keys(suggestions || {}).some((key) => key !== 'fitted_args');
}

export function fitRecommendation(applied, suggestions) {
  const context = applied.ctx_size ? `${formatNumber(applied.ctx_size)} tokens` : 'the fitted context';
  const layers = hasOwn(applied, 'gpu_layers')
    ? (Number(applied.gpu_layers) >= 999 ? 'all GPU layers' : `${formatNumber(applied.gpu_layers)} GPU layers`)
    : 'the fitted GPU layer count';
  const cache = applied.cache_type_k || applied.cache_type_v
    ? `KV cache ${applied.cache_type_k || '-'} / ${applied.cache_type_v || '-'}`
    : 'the fitted KV cache';
  const headroom = suggestions.headroom_mib || applied.fit_headroom_mib;
  if (headroom) {
    return `Use ${context}, ${layers}, and ${cache}. The fit estimates about ${formatNumber(headroom)} MiB of VRAM headroom after load.`;
  }
  return `Use ${context}, ${layers}, and ${cache}. Re-run with a target headroom if you want a stricter VRAM margin.`;
}

export function renderFitSummary(result, applied) {
  const suggestions = result.suggestions || {};
  if (!parsedFitAccepted(suggestions)) {
    const raw = escapeHtml((result.stdout || result.stderr || 'No structured fit suggestion could be parsed.').slice(0, 1800));
    return `<strong>Fit test finished, but no structured recommendation was found.</strong>\n${raw}`;
  }

  const memory = suggestions.cuda_memory_mib || applied.fit_cuda_memory_mib || {};
  const target = applied.fit_target_mib;
  const headroom = suggestions.headroom_mib || applied.fit_headroom_mib;
  const samplingItems = [
    fitItem('Temperature', applied.temperature),
    fitItem('Max tokens', applied.n_predict),
    fitItem('Top K', applied.top_k),
    fitItem('Top P', applied.top_p),
    fitItem('Min P', applied.min_p),
    fitItem('Repeat N', applied.repeat_last_n),
    fitItem('Repeat penalty', applied.repeat_penalty),
    fitItem('Presence penalty', applied.presence_penalty),
    fitItem('Frequency penalty', applied.frequency_penalty),
    fitItem('Seed', applied.seed),
  ].filter(Boolean).join('');
  const memoryItems = [
    fitItem('Context', applied.ctx_size, ' tokens'),
    fitItem('Threads', applied.threads),
    fitItem('Batch threads', applied.threads_batch),
    fitItem('Batch', applied.batch_size),
    fitItem('UBatch', applied.ubatch_size),
    fitItem('GPU layers', Number(applied.gpu_layers) >= 999 ? 'all' : applied.gpu_layers),
    fitItem('Cache K', applied.cache_type_k),
    fitItem('Cache V', applied.cache_type_v),
    fitItem('Target headroom', target, ' MiB'),
    fitItem('Estimated headroom', headroom, ' MiB'),
  ].filter(Boolean).join('');
  const cudaItems = [
    fitItem('Model memory', memory.model, ' MiB'),
    fitItem('Context memory', memory.context, ' MiB'),
    fitItem('Compute memory', memory.compute, ' MiB'),
    fitItem('Projected total', memory.projected, ' MiB'),
  ].filter(Boolean).join('');
  const offloadItems = [
    fitItem('KV cache offload', applied.kv_offload),
    fitItem('CPU helper offload', applied.op_offload),
    fitItem('Flash attention', applied.flash_attn),
    fitItem('Jinja template', applied.jinja),
  ].filter(Boolean).join('');
  const speed = result.speed_estimate;
  const speedItems = speed ? [
    fitItem('Estimated speed', speed.estimate_tps, ' tok/s'),
    fitItem('Likely range', `${speed.low_tps}-${speed.high_tps}`, ' tok/s'),
    fitItem('Confidence', speed.confidence),
  ].filter(Boolean).join('') : '';

  return `
    <div class="fit-summary">
      <div class="fit-status">
        <span class="badge ok">Applied</span>
        <strong>Fit recommendation accepted</strong>
      </div>
      <p>${escapeHtml(fitRecommendation(applied, suggestions))}</p>
      <div class="fit-groups">
        <section>
          <h4>Launch settings</h4>
          <ul>${memoryItems}</ul>
        </section>
        <section>
          <h4>Sampling defaults</h4>
          <ul>${samplingItems}</ul>
        </section>
        ${offloadItems ? `<section><h4>Offload toggles</h4><ul>${offloadItems}</ul></section>` : ''}
        ${cudaItems ? `<section><h4>CUDA estimate</h4><ul>${cudaItems}</ul></section>` : ''}
        ${speedItems ? `<section><h4>Speed estimate</h4><ul>${speedItems}</ul></section>` : ''}
      </div>
      ${suggestions.fitted_args ? `<details class="fit-details"><summary>Fitted CLI args</summary><code>${escapeHtml(suggestions.fitted_args)}</code></details>` : ''}
    </div>
  `;
}

export function fitSummaryText(applied, suggestions, speedEstimate) {
  const parts = [
    'Fit recommendation accepted.',
    `Context: ${applied.ctx_size || '-'}`,
    `Threads: ${applied.threads || '-'} / batch ${applied.threads_batch || applied.threads || '-'}`,
    `Batch: ${applied.batch_size || '-'} / ubatch ${applied.ubatch_size || '-'}`,
    `GPU layers: ${Number(applied.gpu_layers) >= 999 ? 'all' : applied.gpu_layers}`,
    `KV cache: K ${applied.cache_type_k || '-'} / V ${applied.cache_type_v || '-'}`,
    `Temperature: ${applied.temperature ?? '-'}`,
  ];
  const headroom = suggestions.headroom_mib || applied.fit_headroom_mib;
  if (headroom) parts.push(`Estimated headroom: ${headroom} MiB`);
  if (speedEstimate?.estimate_tps) parts.push(`Estimated speed: ${speedEstimate.estimate_tps} tok/s`);
  return parts.join('\n');
}

export async function runFitTest() {
  const mode = selectedMode();
  if (!mode) return;
  const trigger = $('#fit-button');
  setSelectedProfileMode(mode);
  const overrides = saveCurrentOverrides();
  const target = Number($('#param-fit-target').value || 1024);
  setModelNote('fit', 'Running fit test. This may take a moment...');
  await withBusy(trigger, async () => {
    try {
      const result = await api('/api/profiles/fit', {
        method: 'POST',
        body: JSON.stringify({ mode, overrides, target_mib: target, timeout_seconds: 180 }),
      });
      const applied = applyFitResultParams(result);
      const suggestions = result.suggestions || {};
      setModelNote('fit', renderFitSummary(result, applied));
      renderTpsEstimate(result.speed_estimate);
      scheduleTpsEstimate(80);
      showLogPreview(parsedFitAccepted(suggestions)
        ? fitSummaryText(applied, suggestions, result.speed_estimate)
        : (result.command_line || result.stdout || 'Fit test completed.'));
      toast('Fit test complete');
    } catch (error) {
      setModelNote('fit', `<strong>Fit test failed</strong>\n${escapeHtml(error.message)}`);
      toast(`Fit test failed: ${error.message}`);
    }
  });
}

export function applyTuneSuggestion(index) {
  const suggestion = (state.tuneSuggestions || [])[index];
  if (!suggestion) return;
  applyTunedParams(suggestion.params);
  renderTpsEstimate(suggestion.speed_estimate);
  scheduleTpsEstimate(80);
  toast(`Applied ${suggestion.label || 'suggestion'}`);
}

export function applyTunedParams(tuned) {
  const mode = selectedMode();
  if (!mode) return {};
  const applied = { ...collectOverrides(), ...(tuned || {}) };
  setParamOverrides(mode, applied);
  renderParameters();
  markAppliedFields(tuned || {});
  return applied;
}

export async function runAutoTune() {
  const mode = selectedMode();
  if (!mode) return;
  setSelectedProfileMode(mode);
  const overrides = saveCurrentOverrides();
  const trigger = $('#smart-fit-button');
  setModelNote('tune', 'Searching for the best memory fit...');
  await withBusy(trigger, async () => {
    try {
      const result = await api('/api/profiles/auto-tune', {
        method: 'POST',
        body: JSON.stringify({ mode, overrides }),
      });
      if (!result.success) {
        setModelNote('tune', `<strong>Could not auto-tune</strong>\n${escapeHtml(result.reason || 'No fitting configuration found.')}`);
        toast('Smart fit found no safe configuration');
        return;
      }
      state.tuneSuggestions = result.suggestions || [];
      state.jinjaRecommended = !!(result.jinja && result.jinja.recommended);
      const applied = shouldAutoApplyTune(result);
      if (applied) {
        applyTunedParams(result.tuned_params);
        renderTpsEstimate(result.after?.speed_estimate);
        scheduleTpsEstimate(80);
      }
      setModelNote('tune', renderTuneSummary(result, { applied }));
      if (result.cpu_fallback) toast('Smart fit held back a CPU recommendation');
      else toast(applied ? 'Smart fit applied' : 'Smart fit ready');
    } catch (error) {
      setModelNote('tune', `<strong>Smart fit failed</strong>\n${escapeHtml(error.message)}`);
      toast(`Smart fit failed: ${error.message}`);
    }
  });
}

export async function loadSamplingPresets() {
  const select = $('#sampling-intent');
  if (!select) return;
  try {
    const data = await api('/api/sampling/presets');
    state.samplingPresets = data.presets || {};
    select.innerHTML = (data.intents || []).map((intent) => (
      `<option value="${escapeHtml(intent.key)}" title="${escapeHtml(intent.description)}">${escapeHtml(intent.label)}</option>`
    )).join('');
  } catch (error) {
    select.innerHTML = '<option value="">Presets unavailable</option>';
  }
}

export function applySamplingPreset() {
  const mode = selectedMode();
  if (!mode) return;
  const intent = $('#sampling-intent')?.value;
  const preset = state.samplingPresets?.[intent];
  if (!preset?.success) {
    toast('Choose a sampling preset first');
    return;
  }
  setSelectedProfileMode(mode);
  setParamOverrides(mode, { ...saveCurrentOverrides(), ...preset.params });
  renderParameters();
  markAppliedFields(preset.params);
  const rationale = Object.entries(preset.rationale || {})
    .map(([field, why]) => `<li><strong>${escapeHtml(field)}:</strong> ${escapeHtml(why)}</li>`)
    .join('');
  const paramItems = [
    fitItem('Temperature', preset.params.temperature),
    fitItem('Top K', preset.params.top_k),
    fitItem('Top P', preset.params.top_p),
    fitItem('Min P', preset.params.min_p),
    fitItem('Repeat penalty', preset.params.repeat_penalty),
  ].filter(Boolean).join('');
  setModelNote('sampling', `
    <div class="fit-summary">
      <div class="fit-status"><span class="badge ok">Applied</span><strong>${escapeHtml(preset.label)}</strong></div>
      <p>${escapeHtml(preset.description)}</p>
      <div class="fit-groups"><section><h4>Sampling</h4><ul>${paramItems}</ul></section></div>
      ${rationale ? `<details class="fit-details"><summary>Why these values</summary><ul>${rationale}</ul></details>` : ''}
    </div>`);
  toast(`Applied ${preset.label} sampling`);
}

export function renderBenchmarkSummary(result) {
  const benchmark = result.benchmark || {};
  const preview = result.response_preview ? `\n\n${escapeHtml(result.response_preview)}` : '';
  return `
    <div class="fit-summary">
      <div class="fit-status">
        <span class="badge ok">Measured</span>
        <strong>Benchmark complete</strong>
      </div>
      <div class="fit-groups">
        <section>
          <h4>Decode speed</h4>
          <ul>
            ${fitItem('Tokens/sec', benchmark.tokens_per_second)}
            ${fitItem('Generated tokens', benchmark.completion_tokens)}
            ${fitItem('Elapsed', benchmark.elapsed_seconds, ' sec')}
          </ul>
        </section>
        <section>
          <h4>Request</h4>
          <ul>
            ${fitItem('Prompt tokens', benchmark.prompt_tokens)}
            ${fitItem('Total tokens', benchmark.total_tokens)}
            ${fitItem('Chars/sec', benchmark.chars_per_second)}
          </ul>
        </section>
      </div>
      <details class="fit-details"><summary>Response preview</summary><code>${preview || 'No preview returned.'}</code></details>
    </div>
  `;
}

export function renderBenchmarkHistory() {
  const container = $('#benchmark-history');
  const list = $('#benchmark-history-list');
  if (!container || !list) return;

  const all = state.benchmarks || [];
  const currentMode = selectedMode();
  const recent = all.slice(-5).reverse(); // newest first

  if (!recent.length) {
    container.style.display = 'none';
    return;
  }
  container.style.display = '';

  let html = '<table><thead><tr><th>When</th><th>Mode</th><th>t/s</th><th>tok</th><th>Action</th></tr></thead><tbody>';
  recent.forEach((b, i) => {
    const bm = b.benchmark || b;
    const when = b.timestamp ? new Date(b.timestamp).toLocaleTimeString([], {hour:'2-digit', minute:'2-digit'}) : '–';
    const mode = b.mode || '–';
    const tps = bm.tokens_per_second ? bm.tokens_per_second.toFixed(1) : '–';
    const toks = bm.completion_tokens || bm.total_tokens || '–';
    const isCurrent = currentMode && mode === currentMode;
    const cls = isCurrent ? 'current' : '';
    html += `<tr class="${cls}"><td>${escapeHtml(when)}</td><td>${escapeHtml(mode)}</td><td>${tps}</td><td>${toks}</td>`;
    html += `<td><button class="mini-button" data-bench-idx="${all.length - 1 - i}">Use</button></td></tr>`;
  });
  html += '</tbody></table>';
  list.innerHTML = html;

  list.querySelectorAll('button[data-bench-idx]').forEach(btn => {
    btn.addEventListener('click', () => {
      const idx = parseInt(btn.dataset.benchIdx);
      const entry = all[idx];
      if (entry && entry.mode) {
        setSelectedProfileMode(entry.mode);
        renderParameters();
        renderProfiles();
        toast('Loaded params from benchmark history (approximate)');
      }
    });
  });
}

export async function runBenchmark() {
  const mode = selectedMode();
  if (!mode) return;
  setSelectedProfileMode(mode);
  const profile = getSelectedProfile();
  if (!profile?.launchable) {
    toast('Choose a launchable profile before benchmarking');
    return;
  }
  const confirmed = await confirmAction({
    title: 'Run benchmark',
    message: `Benchmark "${profileLabel(mode)}" with the current parameters? This may restart the tracked server for this profile.`,
    confirmLabel: 'Benchmark',
    confirmKind: 'primary',
  });
  if (!confirmed) return;
  const trigger = $('#benchmark-button');
  const overrides = saveCurrentOverrides();
  const requested = Number($('#param-predict').value || 128);
  const completionTokens = requested > 0 ? Math.min(requested, 512) : 128;
  setModelNote('benchmark', 'Running benchmark with the current parameters...');
  await withBusy(trigger, async () => {
    try {
      const result = await api('/api/benchmarks/run', {
        method: 'POST',
        body: JSON.stringify({
          mode,
          overrides,
          completion_tokens: completionTokens,
          restart: true,
          stop_after: false,
          ready_timeout_seconds: 90,
        }),
      });
      setModelNote('benchmark', renderBenchmarkSummary(result));
      renderMeasuredTps(result.benchmark.tokens_per_second, result.benchmark.elapsed_seconds);
      showLogPreview([
        `Benchmark: ${result.benchmark.tokens_per_second} tok/s`,
        `Elapsed: ${result.benchmark.elapsed_seconds}s`,
        `Endpoint: ${result.benchmark.endpoint}`,
      ].join('\n'));
      toast(`Benchmark: ${result.benchmark.tokens_per_second} tok/s`);
      await refresh();
      renderBenchmarkHistory();
      const currentKey = estimateKey(selectedMode() || '', collectOverrides());
      if (currentKey === state.lastBenchmarkKey && state.measuredTps) {
        renderMeasuredTps(state.measuredTps, state.measuredElapsed);
      }
    } catch (error) {
      setModelNote('benchmark', `<strong>Benchmark failed</strong>\n${escapeHtml(error.message)}`);
      toast(`Benchmark failed: ${error.message}`);
    }
  });
}

// Event wiring for this panel, moved out of app.js's wireEvents().
export function initFitPanel() {
  $('#smart-fit-button').addEventListener('click', runAutoTune);
  $('#sampling-suggest-button').addEventListener('click', applySamplingPreset);
  $('#fit-button').addEventListener('click', runFitTest);
  $('#benchmark-button').addEventListener('click', runBenchmark);
}
